### PR TITLE
[WIP] add anomaly detector in zouwu 

### PR DESCRIPTION
**Pull request by shane-huang**
- [x] define detector interface

- [x] define major functionality and scenarios it needs to handle.

- [x] add unit tests

- [x] implementations

### CODE REVIEW

**Comment by jason-dai** [comment]
Please provide some descriptions on how the user will use the API

**Comment by shane-huang** [comment]
> Please provide some descriptions on how the user will use the API

@jason-dai added unit tests to show usage.

File: pyzoo/test/zoo/zouwu/test_model_anomaly.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            del_col_index = {0, 1}.difference(set(target_col_indexes)).pop()
+            data = np.delete(dataset, del_col_index, axis=1)
+        else:
+            data = dataset
+        for i in range(len(data) - look_back - 1):
+            X.append(data[i: (i + look_back)])
+            Y.append(data[i + look_back, target_col_indexes])
+        return np.array(X), np.array(Y)
+
+    def test_app(self):
```
Comment (shane-huang) [inline pyzoo/test/zoo/zouwu/test_model_anomaly.py:55]: This is an example full pipeline from forecast to detection

File: pyzoo/test/zoo/zouwu/test_model_anomaly.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+        assert y_test[anomaly_indexes]
+
+        # if user don't have a threshold, he can choose to use estimator to find a threshold first
+        threshold = ThresholdEstimator.fit(y=y_test, yhat=y_predict, ratio=0.01)
+        fitted_anomaly_indexes = ThresholdDetector.detect(y=y_test, yhat=y_predict, threshold=threshold)
+
+        assert y_test[fitted_anomaly_indexes]
+
+    def test_threshold_case1_multivariant(self):
```
Comment (shane-huang) [inline pyzoo/test/zoo/zouwu/test_model_anomaly.py:90]: This is for single distance handling, where distances of all samples shouldn't go beyond a single value

File: pyzoo/test/zoo/zouwu/test_model_anomaly.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        gen_rand_indexes = np.random.choice(sample_num * feature_dim, num_anomaly)
+        y_test[gen_rand_indexes] = 10
+        y_test = y_test.reshape((sample_num, feature_dim))
+
+        anomaly_indexes = ThresholdDetector.detect(y=y_test,
+                                                   yhat=y_pred,
+                                                   threshold=3)
+        assert len(anomaly_indexes) == num_anomaly
+
+    def test_threshold_case4(self):
```
Comment (shane-huang) [inline pyzoo/test/zoo/zouwu/test_model_anomaly.py:108]: This is for min and max threshold without prediction.

File: pyzoo/zoo/zouwu/model/anomaly.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    def _check_all_distance(self, y, yhat):
+        index = []
+        for i in range(y.shape[0]):
+            diff = self.dist_measure.distance(y[i], yhat[i])
+            if diff >= self.threshold:
+                index.append(i)
+        return index
+
+    def _check_per_dim_distance(self, y, yhat, threshold):
+        raise Exception("Does not support check per dim distance")
```
Comment (shane-huang) [inline pyzoo/zoo/zouwu/model/anomaly.py:160]: maybe change to NotImplementedError.

File: pyzoo/zoo/zouwu/model/anomaly.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+    def _check_per_sample_distance(self, y, yhat, threshold):
+        raise Exception("Does not support check per sample distance")
```
Comment (shane-huang) [inline pyzoo/zoo/zouwu/model/anomaly.py:163]: maybe change to NotImplementedError.

File: pyzoo/zoo/zouwu/model/anomaly.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+            elif len(threshold.shape) == 1:
+                self._check_per_sample_distance(y, yhat, threshold)
+            else:
+                raise ValueError(
+                    "threshold shape", str(
+                        threshold.shape), "is not valid")
+        elif isinstance(threshold, tuple) \
+                and len(threshold) == 2 \
+                and threshold[0].shape == y.shape[1:] \
+                and threshold[-1].shape == y.shape[1:]:
```
Comment (jason-dai) [inline pyzoo/zoo/zouwu/model/anomaly.py:144]: check min <= max?

**Comment by dding3** [comment]
Close this PR since it's duplicated with https://github.com/intel-analytics/analytics-zoo/pull/2296

**Pull request closed without merging**